src/main/java/com/mongodb/spark/sql/connector/MongoCatalog.java

```diff
@@ -0,0 +1,422 @@
+import org.apache.spark.sql.catalyst.analysis.NoSuchNamespaceException;
+import org.apache.spark.sql.catalyst.analysis.NoSuchTableException;
+import org.apache.spark.sql.catalyst.analysis.TableAlreadyExistsException;
+import org.apache.spark.sql.connector.catalog.Identifier;
+import org.apache.spark.sql.connector.catalog.NamespaceChange;
+import org.apache.spark.sql.connector.catalog.SupportsNamespaces;
+import org.apache.spark.sql.connector.catalog.Table;
+import org.apache.spark.sql.connector.catalog.TableCatalog;
+import org.apache.spark.sql.connector.catalog.TableChange;
+import org.apache.spark.sql.connector.expressions.Transform;
+import org.apache.spark.sql.types.StructType;
+import org.apache.spark.sql.util.CaseInsensitiveStringMap;
+import org.jetbrains.annotations.VisibleForTesting;
+
+import com.mongodb.MongoNamespace;
+import com.mongodb.client.MongoCollection;
+import com.mongodb.client.MongoDatabase;
+import com.mongodb.client.model.Filters;
+
+import com.mongodb.spark.sql.connector.connection.MongoConnectionProvider;
+
+/** Spark Catalog methods for working with namespaces (databases) and tables (collections). */
+public class MongoCatalog implements TableCatalog, SupportsNamespaces {
+  private static final String SYSTEM_NAMESPACE_PREFIX = "system.";
+
+  private static boolean excludeSystemNamespaces(final String namespace) {
+    return !namespace.startsWith(SYSTEM_NAMESPACE_PREFIX);
+  }
+
+  private boolean initialized;
+  private String catalogName;
+  private Map<String, String> catalogOptions;
+  private MongoConnectionProvider mongoConnectionProvider;
+
+  /**
+   * Called to initialize configuration.
+   *
+   * <p>This method is called once, just after the provider is instantiated.
+   *
+   * @param name the name used to identify and load this catalog
+   * @param options a case-insensitive string map of configuration
+   */
+  @Override
+  public void initialize(final String name, final CaseInsensitiveStringMap options) {
+    assert !initialized : "The MongoCatalog has already been initialized.";
+    initialized = true;
+    catalogName = name;
+    catalogOptions = options;
+    mongoConnectionProvider = new MongoConnectionProvider(options);
+  }
+
+  /**
+   * Called to get this catalog's name.
+   *
+   * <p>This method is only called after {@link #initialize(String, CaseInsensitiveStringMap)} is
+   * called to pass the catalog's name.
+   */
+  @Override
+  public String name() {
+    assertInitialized();
+    return catalogName;
+  }
+
+  /**
+   * List top-level namespaces (databases) from the catalog.
+   *
+   * <p>MongoDB only has top level namespaces (databases) and does not support nested databases.
+   *
+   * @return an array of multi-part namespace (database) names
+   */
+  @Override
+  public String[][] listNamespaces() {
+    assertInitialized();
+    return mongoConnectionProvider.withClient(
+        client ->
+            client.listDatabaseNames().into(new ArrayList<>()).stream()
```

Comment:
Not sure what the use case for this is, but I wonder if `com.mongodb.client.ListDatabasesIterable#authorizedDatabasesOnly` is relevant.